### PR TITLE
Resource Tagging (#15)

### PR DESCRIPTION
**Pull request by sean-nixon**
Tags are calculated based on a user-provided variable "tags", automatically
calculated Terraform metadata tags (tf-*), and a name_prefix variable used
to calculate the name of resources as well as the "Name" tag. All this is
merged in the final tags applied to resources.

There is an outstanding issue with autoscaling group tagging that makes it not super simple. I'd be open to thoughts from the team, but I think it'll be tricky to get it to work without Terraform 0.12

### CODE REVIEW

**Comment by kulinacs** [comment]
Not sure if I'm doing something wrong, but I'm getting errors attempting to `terraform plan`

```
NMcClendon-MBP[terraform-aws-reference-architectures]% terraform --version
Terraform v0.11.14
+ provider.aws v2.23.0

Your version of Terraform is out of date! The latest version
is 0.12.6. You can update by downloading from www.terraform.io/downloads.html
NMcClendon-MBP[terraform-aws-reference-architectures]% terraform plan

Error: aws_db_instance.master-db: first character of "identifier" must be a letter
```

Changing `name_prefix` to something not starting with a letter, `three-tier` for example, fixes the issue.

File: webtier.tf [hunk 1]
```diff
@@ -39,9 +38,10 @@ resource "aws_lb_listener" "3tier-http-listener" {
 }
 
 resource "aws_security_group" "3tier-lb-sg" {
-  name        = "3tier-lb-sg"
+  name        = "${var.name_prefix}-sg"
   description = "Allow inbound HTTP traffic from everywhere"
   vpc_id      = "${aws_vpc.3tier-vpc.id}"
+  tags        = "${merge(map("Name", "${var.name_prefix}-app-lb-sg"), local.common_tags)}" # TODO Reconsider this for Terraform 0.12
```
Comment (kulinacs) [inline webtier.tf]: Name tag  should be `${var.name_prefix}-lb-sg"`, I believe

Comment (kulinacs) [inline webtier.tf]: Name tag  should be `${var.name_prefix}-lb-sg` or `${var.name_prefix}-web-lb-sg`, I believe

File: variables.tf [hunk 2]
```diff
@@ -119,6 +119,22 @@ variable "app_tier_ami" {
   default     = "ami-024a64a6685d05041"
 }
 
+variable "name_prefix" {
+  description = "Prefix to use when naming resources (also used in the 'Name' tag)"
+  type        = "string"
+  default     = "3tier"
```
Comment (kulinacs) [inline variables.tf]: Mentioned as a comment on the PR itself, setting this to 3tier is causing `terraform plan` to fail for me with:

```
Error: aws_db_instance.master-db: first character of "identifier" must be a letter
```

Comment (sean-nixon) [reply variables.tf]: That's what happens when you make a last minute change to the default thinking it couldn't possibly affect anything and don't test again. Should be fixed now

File: webtier.tf (hunk 1 above)
Comment (sean-nixon) [reply webtier.tf]: Fixed

**Comment by kulinacs** [comment]
lgtm

**Pull request merged**